netboot.c

```diff
@@ -362,8 +372,24 @@ EFI_STATUS FetchNetbootimage(EFI_HANDLE image_handle UNUSED, VOID **buffer, UINT
 		goto try_again;
 	}
 
-	if (EFI_ERROR(efi_status) && *buffer) {
-		FreePool(*buffer);
+	if (EFI_ERROR(efi_status)) {
+		if (pxe->Mode->TftpErrorReceived) {
+			console_print(L"TFTP error %u: %a\n",
+				      pxe->Mode->TftpError.ErrorCode,
+				      pxe->Mode->TftpError.ErrorString);
+
+			switch (pxe->Mode->TftpError.ErrorCode) {
+			case TFTP_ERROR_NOT_FOUND:
```

Comment:
FWIW, I think `TFTP_ERROR_ACCESS` would also make sense to fall back on.. honestly, almost any error from the server side should probably trigger a retry with the default filename.

Same on the HTTP side, 'not found/404' is the obvious one, but 403s, 500s, other things could also be dependent on the filename requested, so it would make sense to retry with the default loader name with almost any error returned from the server.

On the other hand, network layer errors don't make sense to retry with a different filename - but I'm not sure how to differentiate those here.

Reply:
There are really 2 issues here:

1. The shim frontend doesn't handle the errors raised by the network backends. That's what I'm trying to address here. I've chosen to translate the network backend errors to the file backend errors (although I only added file not found here). That way the shim frontend doesn't need to special case the errors returned from the network backends.
2. Shim should fallback to the default loader in more cases. I agree shim should try the default loader for an access denied error. While it's likely reading the default loader will also fail, trying is better than failing outright.

I'm going to rework the patches a bit more so more errors are mapped and then I'll add a commit that also tries the default loader on access denied.

Reply:
I updated the patches to map all TFTP and HTTP error codes to EFI status codes, but I decided not to change the types of errors that shim falls back to the default loader on. The policy that's there is from 50732ee91 and it's to cover the case where shim has incorrectly decided what the second stage loader filename is. If you want the policy to be more like "always try the default second stage if loading the specified one fails", then that's really a separate thing. FWIW, I wrote a patch to add `EFI_ACCESS_DENIED` [here](https://github.com/dbnicholson/shim/commit/dab2c2aa92fb18e27eb35d03c2eb4760970da6cc).